src/Features/LanguageServer/Protocol/RoslynRequestDispatcher.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using CommonLanguageServerProtocol.Framework;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer
+{
+    internal class RoslynRequestDispatcher : RequestDispatcher<RequestContext>, ILspService
+    {
+        public RoslynRequestDispatcher(ILspServices lspServices) : base(lspServices)
+        {
+        }
+
+        private ImmutableDictionary<RequestHandlerMetadata, Lazy<IRequestHandler>>? _requestHandlers;
+
+        public override ImmutableDictionary<RequestHandlerMetadata, Lazy<IRequestHandler>> GetRequestHandlers()
+        {
+            if (_requestHandlers is null)
+            {
+                _requestHandlers = CreateMethodToHandlerMap(_lspServices);
+            }
+
+            return _requestHandlers;
+        }
+
+        private static ImmutableDictionary<RequestHandlerMetadata, Lazy<IRequestHandler>> CreateMethodToHandlerMap(ILspServices lspServices)
```

Comment:
What is keeping this here, and not in the base class?

Reply:
Good call. I was being extra cautious in keeping this out of the base class since not everyone might want to create their handlers via service registration, but if I move this implementation up to the base class and just make it virtual we can give them a good default experience while retaining the ability to override that behavior as needed.